### PR TITLE
feat: enable auth resolution via endpoints

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

upstream: https://github.com/awslabs/smithy-kotlin/pull/993

## Description of changes

Enable using endpoint resolution as a way of resolving auth schemes for select services (S3 and EventBridge). This comes as a requirement from SRA for these services where their auth scheme rules are intimately tied to endpoints. This is an opt-in codegen feature since most services don't need this.

Also refactors the way S3 signing properties are set by defaulting them in the execution context rather than overriding the auth scheme instantiation expression. This ends up being a better customer experience if someone wants to override existing auth schemes (e.g. to replace the signer only) as they don't have to know about all the other S3 specific signer settings.

## Testing

Tested against an MRAP bucket

```kotlin
val sigV4AScheme = SigV4AsymmetricAuthScheme(CrtAwsSigner)

S3Client.fromEnvironment {
    logMode = LogMode.LogRequest
    disableMrap = false
    authSchemes = listOf(sigV4AScheme)
}.use { s3 ->
    s3.putObject {
        bucket = "<redacted>.mrap"
        key = "test-mrap.txt"
        body = ByteStream.fromString("Hello MRAP!")
    }

    s3.listObjectsV2 {
        bucket = "<redacted>.mrap"
    }.contents.orEmpty().forEach(::println)
}

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by lauzadis — APPROVED** [review]

File: codegen/sdk/build.gradle.kts [hunk 1]
```diff
@@ -114,6 +115,20 @@ fun awsServiceProjections(): Provider<List<SmithyProjection>> {
...
+                    node.asObjectNode().get()
+                } else {
+                    Node.objectNode()
+                }
+
+                if (!packageSettings.isEmpty) {
+                    packageSettings.expectMember("sdkId", "${packageSettingsFile.absolutePath} does not contain member `sdkId`")
+                    val packageSdkId = packageSettings.getStringMember("sdkId").get().value
+                    check(service.sdkId == packageSdkId) { "${packageSettingsFile.absolutePath} `sdkId` ($packageSdkId) does not match expected `${service.sdkId}`" }
+                }
```
Comment (ianbotsf) [inline codegen/sdk/build.gradle.kts:130]: **Question**: I'm a little confused by this logic. It looks like if the file doesn't exist or contains only `{}` then we won't assert the existence of an `sdkId`...why is that? Shouldn't it be an error if the file exists and is malformed?

```kotlin
val packageSettingsFile = file(service.packageSettings)
if (packageSettingsFile.exists()) {
    val node = Node.parse(packageSettingsFile.inputStream()).asObjectNode().get()
    node.expectMember("sdkId", "${packageSettingsFile.absolutePath} does not contain member `sdkId`")
    val packageSdkId = node.getStringMember("sdkId").get().value
    check(service.sdkId == packageSdkId) { "${packageSettingsFile.absolutePath} `sdkId` ($packageSdkId) does not match expected `${service.sdkId}`" }
}
```

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt [hunk 2]
```diff
@@ -29,22 +27,14 @@ class S3SigningConfig : KotlinIntegration {
...
-                writer.write("useDoubleUriEncode = false")
-                writer.write("normalizeUriPath = false")
-            }
-        }
+    override val sectionWriters: List<SectionWriterBinding> = listOf(
+        SectionWriterBinding(AwsHttpProtocolClientGenerator.MergeServiceDefaults, ::renderDefaultSigningContext),
+    )
+    private fun renderDefaultSigningContext(writer: KotlinWriter, previousValue: String?) {
+        val signingAttrs = RuntimeTypes.Auth.Signing.AwsSigningCommon.AwsSigningAttributes
+        // https://github.com/awslabs/aws-sdk-kotlin/issues/200
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt:35]: This is here to remind us why path normalization and single URI encoding are important for S3.

**Review by ianbotsf — APPROVED** [review]

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/customization/s3/S3SigningConfig.kt:35]: Yeah I left it as a clue for posterity about why we care about this and where the requirements come from.

File: codegen/sdk/build.gradle.kts [hunk 3]
```diff
@@ -286,6 +304,12 @@ val AwsService.modelExtrasDir: String
 val AwsService.transformsDir: String
     get() = rootProject.file("$destinationDir/transforms").absolutePath
 
+/**
+ * Service specific package settings
+ */
+val AwsService.packageSettings: String
+    get() = rootProject.file("$destinationDir/package.json").absolutePath
```
[the remark this answers is not in the document]
Comment (ianbotsf) [reply codegen/sdk/build.gradle.kts:308]: I don't have a preference on the name—either **package.json** or **config.json** sounds fine to me.

Taking a closer look at this now, does Smithy not already have a way to handle this? If not, should we consider moving this kind of functionality up into the **smithy-kotlin** later?

[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/sdk/build.gradle.kts:308]: Handle what? This is custom logic to feed into `smithy-build.json`.

**Pull request merged**